### PR TITLE
Expose SIGHASH_SINGLE bug in `encode_signing_data_to`

### LINKED ISSUE #1015: Better ways to handle SIGHASH_SINGLE bug
Producing invalid data is surprising and may be difficult to debug.

Ideas for better handling:

* rename the function and return a different result type, renaming should trigger people to check what changed
* return something else than `Result` - weird but may be appropriate in this case
* remove the function - I guess no-go
* try to prevent the situation by construction (didn't look deeply but I guess this is hard/impossible)

Side note: this is a nice example of a bug that would've been prevented by `genio`-like generic IO traits (assuming one would use a proven non-panicking method instead of `expect`).

### PR DESCRIPTION
**Pull request by dpc**
Via `Option` return value

Fix #1015

### CODE REVIEW

File: src/blockdata/transaction.rs [hunk 1]
```diff
@@ -353,56 +377,59 @@ impl Transaction {
...
-                witness: Witness::default(),
-            }];
-        } else {
-            tx.input = Vec::with_capacity(self.input.len());
-            for (n, input) in self.input.iter().enumerate() {
-                tx.input.push(TxIn {
-                    previous_output: input.previous_output,
-                    script_sig: if n == input_index { script_pubkey.clone() } else { Script::new() },
-                    sequence: if n != input_index && (sighash == EcdsaSighashType::Single || sighash == EcdsaSighashType::None) { 0 } else { input.sequence },
+        let mut inner_function = || {
```
Comment (dpc) [inline src/blockdata/transaction.rs:383]: This extra indentation is a bit meh, but is better to make another function, and worry about people calling *that* accidentally? :shrug:

File: src/util/sighash.rs [hunk 2]
```diff
@@ -671,8 +679,11 @@ impl<R: Deref<Target = Transaction>> SighashCache<R> {
         sighash_type: u32,
     ) -> Result<Sighash, Error> {
         let mut enc = Sighash::engine();
-        self.legacy_encode_signing_data_to(&mut enc, input_index, script_pubkey, sighash_type)?;
```
Comment (dpc) [inline src/util/sighash.rs:674]: Was not handling this case a mistake?

**Comment by dpc** [comment]
`Option<Result<(), Error>>` is really a PITA API to handle, so I've added a section explaining it, including example code.

File: src/blockdata/transaction.rs [hunk 3]
```diff
@@ -332,19 +325,50 @@ impl Transaction {
...
+    /// - Does NOT handle the sighash single bug (see "`Option` return value" section)
+    ///
+    /// # `Option` return value
+    ///
+    /// This function can't handle the SIGHASH_SINGLE bug internally. However if it was
+    /// to return an error in this case, a lot of users would mishandle it by calling
+    /// `.expect()` on the result, assuming that `writer` they have passed can't possibly
+    /// fail unaware of this additional case. For this reason the `Result` is wrapped
+    /// in an `Option`. In the case of SIGHASH_SIGNLE bug, the function will return
+    /// `None` without writting anything to the `writer`.
```
Comment (apoelstra) [inline src/blockdata/transaction.rs:337]: In 9435eda1077743065a9e22b031d21275a5ab7f34:

I think it's fine here to just say "we return `None` in the case of the `SIGHASH_SINGLE` bug, reserving the error return of the `Result` for writer errors, allowing users to `unwrap` when they have infallible writers". Also SIGHASH_SINGLE is misspelled once in this block.

File: src/blockdata/transaction.rs [hunk 4]
```diff
@@ -332,19 +325,50 @@ impl Transaction {
...
+    /// # let mut writer = Sighash::engine();
+    /// # let input_index = 0;
+    /// # let script_pubkey = bitcoin::Script::new();
+    /// # let sighash_u32 = 0u32;
+    /// # const SOME_TX: &'static str = "0100000001a15d57094aa7a21a28cb20b59aab8fc7d1149a3bdbcddba9c622e4f5f6a99ece010000006c493046022100f93bb0e7d8db7bd46e40132d1f8242026e045f03a0efe71bbb8e3f475e970d790221009337cd7f1f929f00cc6ff01f03729b069a7c21b59b1736ddfee5db5946c5da8c0121033b9b137ee87d5a812d6f506efdd37f0affa7ffc310711c06c7f3e097c9447c52ffffffff0100e1f505000000001976a9140389035a9225b3839e2bbf32d826a1e222031fd888ac00000000";
+    /// # let raw_tx = Vec::from_hex(SOME_TX).unwrap();
+    /// # let tx: Transaction = deserialize(&raw_tx).unwrap();
+    /// if let None = tx.encode_signing_data_to(&mut writer, input_index, &script_pubkey, sighash_u32)
+    ///         .transpose()
+    ///         .expect("writer doesn't error") {
```
Comment (apoelstra) [inline src/blockdata/transaction.rs:354]: In 9435eda1077743065a9e22b031d21275a5ab7f34:

`.is_none()` is more idiomatic than `if let None =`

File: src/blockdata/transaction.rs (hunk 1 above)
Comment (apoelstra) [reply src/blockdata/transaction.rs:383]: I would slightly prefer just wrapping the whole `{...}` block in `Some(...)`, or alternately defining a new function inside of this one.

Though probably the best solution is to find every return path and add `Some()` individually to each.

File: src/blockdata/transaction.rs (hunk 4 above)
Comment (dpc) [reply src/blockdata/transaction.rs:354]: Ha. :facepalm: :+1:

File: src/blockdata/transaction.rs [hunk 5]
```diff
@@ -353,56 +375,71 @@ impl Transaction {
...
-                witness: Witness::default(),
-            }];
-        } else {
-            tx.input = Vec::with_capacity(self.input.len());
-            for (n, input) in self.input.iter().enumerate() {
-                tx.input.push(TxIn {
-                    previous_output: input.previous_output,
-                    script_sig: if n == input_index { script_pubkey.clone() } else { Script::new() },
-                    sequence: if n != input_index && (sighash == EcdsaSighashType::Single || sighash == EcdsaSighashType::None) { 0 } else { input.sequence },
+        pub fn encode_signing_data_to_inner<Write: io::Write>(
```
Comment (tcharding) [inline src/blockdata/transaction.rs:381]: This can be private.

File: src/blockdata/transaction.rs [hunk 6]
```diff
@@ -353,56 +375,71 @@ impl Transaction {
...
-                output_iter.collect()
-            }
-            EcdsaSighashType::None => vec![],
-            _ => unreachable!()
-        };
-        // hash the result
-        tx.consensus_encode(&mut writer)?;
-        let sighash_arr = endian::u32_to_array_le(sighash_type);
-        sighash_arr.consensus_encode(&mut writer)?;
-        Ok(())
```
Comment (tcharding) [inline src/blockdata/transaction.rs:405]: If the return and these two call sites to `consensus_encode` are changed slightly we can remove the need for the internally defined function.
```rust
        // hash the result
        if let Err(e) = tx.consensus_encode(&mut writer) {
            return Some(Err(e));
        }

        let sighash_arr = endian::u32_to_array_le(sighash_type);
        if let Err(e) = sighash_arr.consensus_encode(&mut writer) {
            return Some(Err(e));
        }
        Some(Ok(()))
```

File: src/util/sighash.rs [hunk 7]
```diff
@@ -36,6 +36,14 @@ use crate::{Script, Transaction, TxOut};
 
 use super::taproot::LeafVersion;
 
+/// Used for signature hash for invalid use of SIGHASH_SINGLE.
+pub(crate) const UINT256_ONE: [u8; 32] = [
```
Comment (tcharding) [inline src/util/sighash.rs:40]: Perhaps this should be defined in the `constants` module and made public as a courtesy because everyone that uses `encode_signing_data_to` is going to have to define this, right? When I defined this array I had to look up what was supposed to be returned, then work out what a "one array" looks like. So, at least for me, it took a little while to do. We could then point users towards it in the function comment.

File: src/util/sighash.rs (hunk 2 above)
Comment (tcharding) [reply src/util/sighash.rs:674]: Looks like it, was probably me that missed it. Good that you caught it.

File: src/blockdata/transaction.rs (hunk 6 above)
Comment (dpc) [reply src/blockdata/transaction.rs:405]: Mixed (not strong) feelings about this one. I kind of like that `?` works.

File: src/util/sighash.rs (hunk 7 above)
Comment (dpc) [reply src/util/sighash.rs:40]: Should I rename it something more meaningful or is the current name OK?

File: src/blockdata/transaction.rs (hunk 6 above)
Comment (tcharding) [reply src/blockdata/transaction.rs:405]: No worries, happy to defer to whatever you prefer.

File: src/util/sighash.rs (hunk 7 above)
Comment (tcharding) [reply src/util/sighash.rs:40]: If you can come up with a better name by all means rename it, that was the best I could do at the time :)

**Review by tcharding — DISMISSED** [review]
ACK c174fc144c23b59292d46c3c8787e4cc9ec1023e

**Comment by dpc** [comment]
The only thing left seems making `UINT256_ONE` public, move to constant and (consider if the name is good). The name didn't matter when it was private, but now I guess is a good time to commit to it. Thoughts?

**Comment by tcharding** [comment]
Its an unsigned integer with value one and the comment on it says what its for so I rekon its right to leave it as it is.

**Comment by tcharding** [comment]
I'm tempted to suggest just leaving this PR as it is and re-visiting the public thing if it gets requested. Its a trivial PR for drive by devs so I would imagine someone would do it if they wanted it.

**Comment by Kixunil** [comment]
Isn't there a problem that `Option` still has `unwrap` method so the existing code unwrapping will continue to panic? Although there should be at least a warning about returned type being `Result` so it's some improvement.

Possible alternative:

```rust
#[must_use]
enum SighashResult {
    Success,
    Error(io::Error),
    SighashSingleBug,
}

impl SighashResult {
    /// Returns false if there's SIGHASH_SINGLE bug, panics if writer actually errored.
    #[must_use]
    fn is_success_writer_never_errors(self) -> bool {
        match self {
            SighashResult::Success => true,
            // lazy POC doesn't display sources
            SighashResult::Error(error) => panic!("writer shouldn't have failed but it did: {}", error),
            SighashResult::SighashSingleBug => bool,
        }
    }

    // no unwrap() method available
    // Maybe other helper methods?
}
```

Then your example would just be:

```rust
let result = tx.encode_signing_data_to(&mut writer, input_index, &script_pubkey, sighash_u32);
if  result.is_success_writer_never_errors() {
    // use a hash value of "1", instead of computing the actual hash due to SIGHASH_SIGNLE bug
}
```

It adds a bunch of code on our side but I'd argue the consumer code would be more readable and robust. Anyway, I'd really love to support genio-style IO with non-erroring writers being proven by the type system.

**Comment by apoelstra** [comment]
@Kixunil I'm not sure how existing code that calls `unwrap` on a `Result` could continue to work on an `Option<Result>` without noticing a compiler error.

File: src/blockdata/transaction.rs [hunk 8]
```diff
@@ -439,8 +466,11 @@ impl Transaction {
         }
 
         let mut engine = Sighash::engine();
-        self.encode_signing_data_to(&mut engine, input_index, script_pubkey, sighash_u32)
-            .expect("engines don't error");
+        if let None = self.encode_signing_data_to(&mut engine, input_index, script_pubkey, sighash_u32)
```
Comment (sanket1729) [inline src/blockdata/transaction.rs]: nit: prefer `is_none` over `if let None`

Comment (sanket1729) [reply src/blockdata/transaction.rs]: https://rust-lang.github.io/rust-clippy/v0.0.177/#if_let_redundant_pattern_matching

File: src/blockdata/transaction.rs [hunk 9]
```diff
@@ -332,19 +325,50 @@ impl Transaction {
...
+    /// # use bitcoin::{Transaction, Sighash};
+    /// # use bitcoin_hashes::{Hash, hex::FromHex};
+    /// # let mut writer = Sighash::engine();
+    /// # let input_index = 0;
+    /// # let script_pubkey = bitcoin::Script::new();
+    /// # let sighash_u32 = 0u32;
+    /// # const SOME_TX: &'static str = "0100000001a15d57094aa7a21a28cb20b59aab8fc7d1149a3bdbcddba9c622e4f5f6a99ece010000006c493046022100f93bb0e7d8db7bd46e40132d1f8242026e045f03a0efe71bbb8e3f475e970d790221009337cd7f1f929f00cc6ff01f03729b069a7c21b59b1736ddfee5db5946c5da8c0121033b9b137ee87d5a812d6f506efdd37f0affa7ffc310711c06c7f3e097c9447c52ffffffff0100e1f505000000001976a9140389035a9225b3839e2bbf32d826a1e222031fd888ac00000000";
+    /// # let raw_tx = Vec::from_hex(SOME_TX).unwrap();
+    /// # let tx: Transaction = deserialize(&raw_tx).unwrap();
+    /// if let None = tx.encode_signing_data_to(&mut writer, input_index, &script_pubkey, sighash_u32)
```
Comment (sanket1729) [inline src/blockdata/transaction.rs]: nit: Same comment about `is_none` here

File: src/util/sighash.rs [hunk 10]
```diff
@@ -671,8 +680,11 @@ impl<R: Deref<Target = Transaction>> SighashCache<R> {
         sighash_type: u32,
     ) -> Result<Sighash, Error> {
         let mut enc = Sighash::engine();
-        self.legacy_encode_signing_data_to(&mut enc, input_index, script_pubkey, sighash_type)?;
-        Ok(Sighash::from_engine(enc))
+        if let None = self.legacy_encode_signing_data_to(&mut enc, input_index, script_pubkey, sighash_type).transpose()? {
```
Comment (sanket1729) [inline src/util/sighash.rs]: nit: Same comment here

**Review by sanket1729 — DISMISSED** [review]
ACK fac608d8379986495baa3f45fc599e26dcd9a748. Looks like my nits were addressed in the next commits. Can you also squash the last few fixup commits.

**Comment by apoelstra** [comment]
@dpc I asked for a description of how a user could write code that `unwrap`s a `Result` such that the same code would continue to compile with an `Option<Result>`. It looks like you demonstrated one, so my :confused: emoji was in response to that, because I was surprised to see this was possible.

Regardnig `SighashResult` existing and/or the set of conversion methods we want, I have no opinion.

**Comment by Kixunil** [comment]
@dpc I think `Result<bool, io::Error>` would be best. If your writer can error, you can just `?`. Something like:

```
let result = tx.encode_signing_data_to(&mut writer, input_index, &script_pubkey, sighash_u32);
if  result.is_sighash_bug()? {
    // use a hash value of "1", instead of computing the actual hash due to SIGHASH_SIGNLE bug
}
```

**Comment by dpc** [comment]
```rust
if  tx.encode_signing_data_to(&mut writer, input_index, &script_pubkey, sighash_u32).is_sighash_bug()? {
    // use a hash value of "1", instead of computing the actual hash due to SIGHASH_SIGNLE bug
}
```

(which is the same, just without a tmp-value).

works with me. I can change it tomorrow if no one is against.  (or feel free to take over, always fine with me)

**Comment by apoelstra** [comment]
I like it!

**Comment by Kixunil** [comment]
@dpc the reason for temporary value is to avoid long `if` line, which I don't like but wouldn't block the PR for.

**Comment by dpc** [comment]
I renamed the type to correspond to the functionality it actually is returned from. I've also made the whole `Result` to be separate variant - it just seem easier and avoids having non-idiomatic `Success` (vs `Ok`) and `Error` (vs `Err`) and confusing this with regular `Result`.

File: src/blockdata/transaction.rs [hunk 11]
```diff
@@ -262,6 +255,58 @@ impl Default for TxOut {
...
+/// This type forces the caller to handle SIGHASH_SINGLE bug corner case.
+pub enum EncodeSigningDataResult<E> {
+    /// Input data is an instance of `SIGHASH_SINGLE` bug
+    SighashSingleBug,
+    /// Operation performed normally.
+    Result(Result<(), E>),
+}
+
+impl<E> EncodeSigningDataResult<E> {
+    /// Caller is recommended to use the following pattern to handle this corner case:
```
Comment (Kixunil) [inline src/blockdata/transaction.rs:269]: A summary would be nice + maybe reword/expand this. Something like:

> Checks for `SIGHASH_SINGLE` bug returning error if the writer failed.
>
> This is method is provided for easy and correct handling of the result because `SIGHASH_SINGLE` bug is a special case that must not be ignored nor cause panicking. Since the data is usually written directly into a hasher which never fails, the recommended pattern to handle this is:

File: src/blockdata/transaction.rs [hunk 12]
```diff
@@ -370,19 +415,24 @@ impl Transaction {
     /// - Does NOT attempt to support OP_CODESEPARATOR. In general this would require evaluating
     /// `script_pubkey` to determine which separators get evaluated and which don't, which we don't
     /// have the information to determine.
-    /// - Does NOT handle the sighash single bug, you should either handle that manually or use
-    /// [`Self::signature_hash()`] instead.
+    /// - Does NOT handle the sighash single bug (see "`Option` return value" section)
```
Comment (Kixunil) [inline src/blockdata/transaction.rs:418]: There's no longer `Option`

File: src/blockdata/transaction.rs [hunk 13]
```diff
@@ -370,19 +415,24 @@ impl Transaction {
...
+    ///
+    /// # Return type
+    ///
+    /// This function can't handle the SIGHASH_SINGLE bug internally, so it returns [`EncodeSigningResult`]
```
Comment (Kixunil) [inline src/blockdata/transaction.rs:422]: Should be `EncodeSigningDataResult`, also maybe recommend checking out `EncodeSigningDataResult::is_sighash_single_bug` for quick handling.

File: src/blockdata/transaction.rs [hunk 14]
```diff
@@ -262,6 +255,58 @@ impl Default for TxOut {
...
 }
 
+/// Result of [`Transaction::encode_signing_data_to`].
+///
+/// This type forces the caller to handle SIGHASH_SINGLE bug corner case.
+pub enum EncodeSigningDataResult<E> {
+    /// Input data is an instance of `SIGHASH_SINGLE` bug
+    SighashSingleBug,
+    /// Operation performed normally.
+    Result(Result<(), E>),
```
Comment (Kixunil) [inline src/blockdata/transaction.rs:265]: I don't like naming custom items with names like `Result`, `Ok` because of confusion. This case is pretty reasonable but consider naming it `WriteResult` or something else. If we can't figure out a better name, it's fine.

File: src/util/sighash.rs [hunk 15]
```diff
@@ -665,8 +675,12 @@ impl<R: Deref<Target = Transaction>> SighashCache<R> {
         sighash_type: u32,
     ) -> Result<Sighash, Error> {
         let mut enc = Sighash::engine();
-        self.legacy_encode_signing_data_to(&mut enc, input_index, script_pubkey, sighash_type)?;
-        Ok(Sighash::from_engine(enc))
+        if self.legacy_encode_signing_data_to(&mut enc, input_index, script_pubkey, sighash_type)
+            .is_sighash_single_bug()? {
+            Ok(Sighash::from_slice(&UINT256_ONE).expect("const-size array"))
```
Comment (Kixunil) [inline src/util/sighash.rs:680]: `from_inner`

File: src/blockdata/transaction.rs [hunk 16]
```diff
@@ -477,8 +544,11 @@ impl Transaction {
         }
 
         let mut engine = Sighash::engine();
-        self.encode_signing_data_to(&mut engine, input_index, script_pubkey, sighash_u32)
-            .expect("engines don't error");
+        if self.encode_signing_data_to(&mut engine, input_index, script_pubkey, sighash_u32)
+            .is_sighash_single_bug()
+            .expect("engines don't error") {
+            return Sighash::from_slice(&UINT256_ONE).expect("const-size array");
```
Comment (Kixunil) [inline src/blockdata/transaction.rs:550]: `Sighash::from_inner(UINT256_ONE)` should work & avoid `expect`. I also wonder if it'd be better to make `UINT256_ONE` and associated const/method of `Sighash` intended for this purpose. Not sure how exactly to name it though.

**Review by Kixunil — COMMENTED** [review]
Mostly doc/style issues. Could you make avoiding monomorphisation a separate commit? It's a bit annoying to review like this, so I didn't yet.

File: src/blockdata/transaction.rs (hunk 16 above)
Comment (dpc) [reply src/blockdata/transaction.rs:550]: The name and place for `UINT256_ONE` are to be figured out in another PR. `constants` were suggested.

File: src/blockdata/transaction.rs [hunk 17]
```diff
@@ -391,56 +441,73 @@ impl Transaction {
...
-                witness: Witness::default(),
-            }];
-        } else {
-            tx.input = Vec::with_capacity(self.input.len());
-            for (n, input) in self.input.iter().enumerate() {
-                tx.input.push(TxIn {
-                    previous_output: input.previous_output,
-                    script_sig: if n == input_index { script_pubkey.clone() } else { Script::new() },
-                    sequence: if n != input_index && (sighash == EcdsaSighashType::Single || sighash == EcdsaSighashType::None) { 0 } else { input.sequence },
+        fn encode_signing_data_to_inner<Write: io::Write>(
```
[the remark this answers is not in the document]
Comment (dpc) [reply src/blockdata/transaction.rs:455]:
```
437 |     pub fn encode_signing_data_to<Write: io::Write, U: Into<u32>>(·
    |                                   ----- type parameter from outer function·
...·
455 |         fn encode_signing_data_to_inner(·
    |            ---------------------------- try adding a local generic parameter in this method instead·
456 |             self_: &Transaction,·
457 |             mut writer: Write,·
    |                         ^^^^^ use of generic parameter from outer function·
```

I'm not sure what you mean, but if that - it doesn't work.

[the remark this answers is not in the document]
Comment (Kixunil) [reply src/blockdata/transaction.rs:455]: The error message suggests it does avoid monomorphisation. :)

File: src/blockdata/transaction.rs [hunk 18]
```diff
@@ -415,17 +424,16 @@ impl Transaction {
...
     /// have the information to determine.
-    /// - Does NOT handle the sighash single bug (see "`Option` return value" section)
+    /// - Does NOT handle the sighash single bug (see "Return type" section)
     ///
     /// # Return type
     ///
-    /// This function can't handle the SIGHASH_SINGLE bug internally, so it returns [`EncodeSigningResult`]
-    /// that must be handled by the caller.
+    /// This function can't handle the SIGHASH_SINGLE bug internally, so it returns [`EncodeSigningDataResult`]
+    /// that must be handled by the caller (see [`EncodeSigningDataResult::is_sighash_single_bug`].
```
Comment (Kixunil) [inline src/blockdata/transaction.rs:432]: Nit: missing closing parenthesis

**Review by Kixunil — COMMENTED** [review]
Great, now just clean up git history and you should get an ACK from me. :)

**Comment by dpc** [comment]
Done.

File: src/blockdata/transaction.rs [hunk 19]
```diff
@@ -262,6 +255,67 @@ impl Default for TxOut {
...
+    /// Input data is an instance of `SIGHASH_SINGLE` bug
+    SighashSingleBug,
+    /// Operation performed normally.
+    WriteResult(Result<(), E>),
+}
+
+impl<E> EncodeSigningDataResult<E> {
+    /// Checks for SIGHASH_SINGLE bug returning error if the writer failed.
+    ///
+    /// This is method is provided for easy and correct handling of the result because
```
Comment (tcharding) [inline src/blockdata/transaction.rs:276]: nit:

```suggestion
    /// This method is provided for easy and correct handling of the result because
```

**Review by tcharding — APPROVED** [review]
ACK 5f9a22d4e40a656798ad00599989f941dfc73fb8

File: src/blockdata/transaction.rs (hunk 19 above)
Comment (tcharding) [reply src/blockdata/transaction.rs:276]: I built and reviewed the HTML docs.

**Comment by tcharding** [comment]
I built and reviewed the rendered HTML docs, except for one minor nit, it all looks super good. Cheers!

**Comment by sanket1729** [comment]
Super for the super late jump-in. Feel free to ignore my suggestion. Instead of exposing the bug, I think we can fix it. What if we added an enum in Result. The function would then return `Result<Status, Error>`
```rust

struct Status {

    SighashSingleMisMatch,

    RegularLegacySignatureHash,
}
```
The slight advantage we have here is we can correctly write the ALLONE value in hash function. In the documentation, we can then explain when the return value is `SighashSingleMismatch` do not hash the value. If there is interest here, feel free to bike-shed the names.

**Review by tcharding — APPROVED** [review]
ACK 42a91ab32ab3b02ebfa2e2dc2dcbdad73b4757a4

**Review by apoelstra — APPROVED** [review]
ACK 42a91ab32ab3b02ebfa2e2dc2dcbdad73b4757a4

**Comment by dpc** [comment]
Merge?
